// angular2/src/forms/directives.js
"use strict";
Object.defineProperties(module.exports, {
  DefaultValueAccessor: {get: function() {
      return DefaultValueAccessor;
    }},
  CheckboxControlValueAccessor: {get: function() {
      return CheckboxControlValueAccessor;
    }},
  ControlDirective: {get: function() {
      return ControlDirective;
    }},
  ControlGroupDirective: {get: function() {
      return ControlGroupDirective;
    }},
  FormDirectives: {get: function() {
      return FormDirectives;
    }},
  __esModule: {value: true}
});
var $__rtts_95_assert_47_rtts_95_assert__,
    $__angular2_47_angular2__,
    $__angular2_47_di__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_collection__,
    $__model__,
    $__validators__;
var assert = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__}).assert;
var $__1 = ($__angular2_47_angular2__ = require("angular2/angular2"), $__angular2_47_angular2__ && $__angular2_47_angular2__.__esModule && $__angular2_47_angular2__ || {default: $__angular2_47_angular2__}),
    Template = $__1.Template,
    Component = $__1.Component,
    Decorator = $__1.Decorator,
    Ancestor = $__1.Ancestor,
    onChange = $__1.onChange,
    PropertySetter = $__1.PropertySetter;
var Optional = ($__angular2_47_di__ = require("angular2/di"), $__angular2_47_di__ && $__angular2_47_di__.__esModule && $__angular2_47_di__ || {default: $__angular2_47_di__}).Optional;
var $__3 = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}),
    isBlank = $__3.isBlank,
    isPresent = $__3.isPresent,
    isString = $__3.isString,
    CONST = $__3.CONST;
var $__4 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    StringMapWrapper = $__4.StringMapWrapper,
    ListWrapper = $__4.ListWrapper;
var $__5 = ($__model__ = require("./model"), $__model__ && $__model__.__esModule && $__model__ || {default: $__model__}),
    ControlGroup = $__5.ControlGroup,
    Control = $__5.Control;
var Validators = ($__validators__ = require("./validators"), $__validators__ && $__validators__.__esModule && $__validators__ || {default: $__validators__}).Validators;
var DefaultValueAccessor = function DefaultValueAccessor(setValueProperty) {
  assert.argumentTypes(setValueProperty, Function);
  this._setValueProperty = setValueProperty;
  this.onChange = (function(_) {});
};
($traceurRuntime.createClass)(DefaultValueAccessor, {writeValue: function(value) {
    this._setValueProperty(value);
  }}, {});
Object.defineProperty(DefaultValueAccessor, "annotations", {get: function() {
    return [new Decorator({
      selector: '[control]',
      events: {
        'change': 'onChange($event.target.value)',
        'input': 'onChange($event.target.value)'
      }
    })];
  }});
Object.defineProperty(DefaultValueAccessor, "parameters", {get: function() {
    return [[Function, new PropertySetter('value')]];
  }});
var CheckboxControlValueAccessor = function CheckboxControlValueAccessor(cd, setCheckedProperty) {
  assert.argumentTypes(cd, ControlDirective, setCheckedProperty, Function);
  this._setCheckedProperty = setCheckedProperty;
  this.onChange = (function(_) {});
  cd.valueAccessor = this;
};
($traceurRuntime.createClass)(CheckboxControlValueAccessor, {writeValue: function(value) {
    this._setCheckedProperty(value);
  }}, {});
Object.defineProperty(CheckboxControlValueAccessor, "annotations", {get: function() {
    return [new Decorator({
      selector: 'input[type=checkbox]',
      events: {'change': 'onChange($event.target.checked)'}
    })];
  }});
Object.defineProperty(CheckboxControlValueAccessor, "parameters", {get: function() {
    return [[ControlDirective], [Function, new PropertySetter('checked')]];
  }});
var ControlDirective = function ControlDirective(groupDirective, valueAccessor) {
  assert.argumentTypes(groupDirective, ControlGroupDirective, valueAccessor, DefaultValueAccessor);
  this._groupDirective = groupDirective;
  this.controlOrName = null;
  this.valueAccessor = valueAccessor;
  this.validator = Validators.nullValidator;
};
($traceurRuntime.createClass)(ControlDirective, {
  onChange: function(_) {
    this._initialize();
  },
  _initialize: function() {
    if (isPresent(this._groupDirective)) {
      this._groupDirective.addDirective(this);
    }
    var c = this._control();
    c.validator = Validators.compose([c.validator, this.validator]);
    this._updateDomValue();
    this._setUpUpdateControlValue();
  },
  _updateDomValue: function() {
    this.valueAccessor.writeValue(this._control().value);
  },
  _setUpUpdateControlValue: function() {
    var $__7 = this;
    this.valueAccessor.onChange = (function(newValue) {
      return $__7._control().updateValue(newValue);
    });
  },
  _control: function() {
    if (isString(this.controlOrName)) {
      return this._groupDirective.findControl(this.controlOrName);
    } else {
      return this.controlOrName;
    }
  }
}, {});
Object.defineProperty(ControlDirective, "annotations", {get: function() {
    return [new Decorator({
      lifecycle: [onChange],
      selector: '[control]',
      bind: {'controlOrName': 'control'}
    })];
  }});
Object.defineProperty(ControlDirective, "parameters", {get: function() {
    return [[ControlGroupDirective, new Optional(), new Ancestor()], [DefaultValueAccessor]];
  }});
var ControlGroupDirective = function ControlGroupDirective(groupDirective) {
  assert.argumentTypes(groupDirective, $ControlGroupDirective);
  this._groupDirective = groupDirective;
  this._directives = ListWrapper.create();
};
var $ControlGroupDirective = ControlGroupDirective;
($traceurRuntime.createClass)(ControlGroupDirective, {
  set controlGroup(controlGroup) {
    if (isString(controlGroup)) {
      this._controlGroupName = controlGroup;
    } else {
      this._controlGroup = controlGroup;
    }
    this._updateDomValue();
  },
  _updateDomValue: function() {
    ListWrapper.forEach(this._directives, (function(cd) {
      return cd._updateDomValue();
    }));
  },
  addDirective: function(c) {
    assert.argumentTypes(c, ControlDirective);
    ListWrapper.push(this._directives, c);
  },
  findControl: function(name) {
    assert.argumentTypes(name, assert.type.string);
    return assert.returnType((this._getControlGroup().controls[name]), assert.type.any);
  },
  _getControlGroup: function() {
    if (isPresent(this._controlGroupName)) {
      return assert.returnType((this._groupDirective.findControl(this._controlGroupName)), ControlGroup);
    } else {
      return assert.returnType((this._controlGroup), ControlGroup);
    }
  }
}, {});
Object.defineProperty(ControlGroupDirective, "annotations", {get: function() {
    return [new Decorator({
      selector: '[control-group]',
      bind: {'controlGroup': 'control-group'}
    })];
  }});
Object.defineProperty(ControlGroupDirective, "parameters", {get: function() {
    return [[ControlGroupDirective, new Optional(), new Ancestor()]];
  }});
Object.defineProperty(ControlGroupDirective.prototype.addDirective, "parameters", {get: function() {
    return [[ControlDirective]];
  }});
Object.defineProperty(ControlGroupDirective.prototype.findControl, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
var FormDirectives = [ControlGroupDirective, ControlDirective, CheckboxControlValueAccessor, DefaultValueAccessor];
