// angular2/src/forms/model.js
"use strict";
Object.defineProperties(module.exports, {
  VALID: {get: function() {
      return VALID;
    }},
  INVALID: {get: function() {
      return INVALID;
    }},
  AbstractControl: {get: function() {
      return AbstractControl;
    }},
  Control: {get: function() {
      return Control;
    }},
  ControlGroup: {get: function() {
      return ControlGroup;
    }},
  ControlArray: {get: function() {
      return ControlArray;
    }},
  __esModule: {value: true}
});
var $__rtts_95_assert_47_rtts_95_assert__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_async__,
    $__angular2_47_src_47_facade_47_collection__,
    $__validators__;
var assert = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__}).assert;
var isPresent = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}).isPresent;
var $__2 = ($__angular2_47_src_47_facade_47_async__ = require("angular2/src/facade/async"), $__angular2_47_src_47_facade_47_async__ && $__angular2_47_src_47_facade_47_async__.__esModule && $__angular2_47_src_47_facade_47_async__ || {default: $__angular2_47_src_47_facade_47_async__}),
    Observable = $__2.Observable,
    ObservableController = $__2.ObservableController,
    ObservableWrapper = $__2.ObservableWrapper;
var $__3 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    StringMap = $__3.StringMap,
    StringMapWrapper = $__3.StringMapWrapper,
    ListWrapper = $__3.ListWrapper,
    List = $__3.List;
var Validators = ($__validators__ = require("./validators"), $__validators__ && $__validators__.__esModule && $__validators__ || {default: $__validators__}).Validators;
var VALID = "VALID";
var INVALID = "INVALID";
var AbstractControl = function AbstractControl(validator) {
  assert.argumentTypes(validator, Function);
  this.validator = validator;
  this._pristine = true;
};
($traceurRuntime.createClass)(AbstractControl, {
  get value() {
    return assert.returnType((this._value), assert.type.any);
  },
  get status() {
    return assert.returnType((this._status), assert.type.string);
  },
  get valid() {
    return assert.returnType((this._status === VALID), assert.type.boolean);
  },
  get errors() {
    return assert.returnType((this._errors), StringMap);
  },
  get pristine() {
    return assert.returnType((this._pristine), assert.type.boolean);
  },
  get dirty() {
    return assert.returnType((!this.pristine), assert.type.boolean);
  },
  setParent: function(parent) {
    this._parent = parent;
  },
  _updateParent: function() {
    if (isPresent(this._parent)) {
      this._parent._updateValue();
    }
  }
}, {});
Object.defineProperty(AbstractControl, "parameters", {get: function() {
    return [[Function]];
  }});
var Control = function Control(value) {
  var validator = arguments[1] !== (void 0) ? arguments[1] : Validators.nullValidator;
  assert.argumentTypes(value, assert.type.any, validator, Function);
  $traceurRuntime.superConstructor($Control).call(this, validator);
  this._setValueErrorsStatus(value);
  this._valueChangesController = ObservableWrapper.createController();
  this.valueChanges = ObservableWrapper.createObservable(this._valueChangesController);
};
var $Control = Control;
($traceurRuntime.createClass)(Control, {
  updateValue: function(value) {
    assert.argumentTypes(value, assert.type.any);
    this._setValueErrorsStatus(value);
    this._pristine = false;
    ObservableWrapper.callNext(this._valueChangesController, this._value);
    this._updateParent();
  },
  _setValueErrorsStatus: function(value) {
    this._value = value;
    this._errors = this.validator(this);
    this._status = isPresent(this._errors) ? INVALID : VALID;
  }
}, {}, AbstractControl);
Object.defineProperty(Control, "parameters", {get: function() {
    return [[assert.type.any], [Function]];
  }});
Object.defineProperty(Control.prototype.updateValue, "parameters", {get: function() {
    return [[assert.type.any]];
  }});
var ControlGroup = function ControlGroup(controls) {
  var optionals = arguments[1] !== (void 0) ? arguments[1] : null;
  var validator = arguments[2] !== (void 0) ? arguments[2] : Validators.group;
  assert.argumentTypes(controls, StringMap, optionals, StringMap, validator, Function);
  $traceurRuntime.superConstructor($ControlGroup).call(this, validator);
  this.controls = controls;
  this._optionals = isPresent(optionals) ? optionals : {};
  this._valueChangesController = ObservableWrapper.createController();
  this.valueChanges = ObservableWrapper.createObservable(this._valueChangesController);
  this._setParentForControls();
  this._setValueErrorsStatus();
};
var $ControlGroup = ControlGroup;
($traceurRuntime.createClass)(ControlGroup, {
  include: function(controlName) {
    assert.argumentTypes(controlName, assert.type.string);
    StringMapWrapper.set(this._optionals, controlName, true);
    this._updateValue();
  },
  exclude: function(controlName) {
    assert.argumentTypes(controlName, assert.type.string);
    StringMapWrapper.set(this._optionals, controlName, false);
    this._updateValue();
  },
  contains: function(controlName) {
    assert.argumentTypes(controlName, assert.type.string);
    var c = StringMapWrapper.contains(this.controls, controlName);
    return assert.returnType((c && this._included(controlName)), assert.type.boolean);
  },
  _setParentForControls: function() {
    var $__5 = this;
    StringMapWrapper.forEach(this.controls, (function(control, name) {
      control.setParent($__5);
    }));
  },
  _updateValue: function() {
    this._setValueErrorsStatus();
    this._pristine = false;
    ObservableWrapper.callNext(this._valueChangesController, this._value);
    this._updateParent();
  },
  _setValueErrorsStatus: function() {
    this._value = this._reduceValue();
    this._errors = this.validator(this);
    this._status = isPresent(this._errors) ? INVALID : VALID;
  },
  _reduceValue: function() {
    return this._reduceChildren({}, (function(acc, control, name) {
      acc[name] = control.value;
      return acc;
    }));
  },
  _reduceChildren: function(initValue, fn) {
    var $__5 = this;
    assert.argumentTypes(initValue, assert.type.any, fn, Function);
    var res = initValue;
    StringMapWrapper.forEach(this.controls, (function(control, name) {
      if ($__5._included(name)) {
        res = fn(res, control, name);
      }
    }));
    return res;
  },
  _included: function(controlName) {
    assert.argumentTypes(controlName, assert.type.string);
    var isOptional = StringMapWrapper.contains(this._optionals, controlName);
    return assert.returnType((!isOptional || StringMapWrapper.get(this._optionals, controlName)), assert.type.boolean);
  }
}, {}, AbstractControl);
Object.defineProperty(ControlGroup, "parameters", {get: function() {
    return [[StringMap], [StringMap], [Function]];
  }});
Object.defineProperty(ControlGroup.prototype.include, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
Object.defineProperty(ControlGroup.prototype.exclude, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
Object.defineProperty(ControlGroup.prototype.contains, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
Object.defineProperty(ControlGroup.prototype._reduceChildren, "parameters", {get: function() {
    return [[assert.type.any], [Function]];
  }});
Object.defineProperty(ControlGroup.prototype._included, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
var ControlArray = function ControlArray(controls) {
  var validator = arguments[1] !== (void 0) ? arguments[1] : Validators.array;
  assert.argumentTypes(controls, assert.genericType(List, AbstractControl), validator, Function);
  $traceurRuntime.superConstructor($ControlArray).call(this, validator);
  this.controls = controls;
  this._valueChangesController = ObservableWrapper.createController();
  this.valueChanges = ObservableWrapper.createObservable(this._valueChangesController);
  this._setParentForControls();
  this._setValueErrorsStatus();
};
var $ControlArray = ControlArray;
($traceurRuntime.createClass)(ControlArray, {
  at: function(index) {
    assert.argumentTypes(index, assert.type.number);
    return assert.returnType((this.controls[index]), AbstractControl);
  },
  push: function(control) {
    assert.argumentTypes(control, AbstractControl);
    ListWrapper.push(this.controls, control);
    control.setParent(this);
    this._updateValue();
  },
  insert: function(index, control) {
    assert.argumentTypes(index, assert.type.number, control, AbstractControl);
    ListWrapper.insert(this.controls, index, control);
    control.setParent(this);
    this._updateValue();
  },
  removeAt: function(index) {
    assert.argumentTypes(index, assert.type.number);
    ListWrapper.removeAt(this.controls, index);
    this._updateValue();
  },
  get length() {
    return assert.returnType((this.controls.length), assert.type.number);
  },
  _updateValue: function() {
    this._setValueErrorsStatus();
    this._pristine = false;
    ObservableWrapper.callNext(this._valueChangesController, this._value);
    this._updateParent();
  },
  _setParentForControls: function() {
    var $__5 = this;
    ListWrapper.forEach(this.controls, (function(control) {
      control.setParent($__5);
    }));
  },
  _setValueErrorsStatus: function() {
    this._value = ListWrapper.map(this.controls, (function(c) {
      return c.value;
    }));
    this._errors = this.validator(this);
    this._status = isPresent(this._errors) ? INVALID : VALID;
  }
}, {}, AbstractControl);
Object.defineProperty(ControlArray, "parameters", {get: function() {
    return [[assert.genericType(List, AbstractControl)], [Function]];
  }});
Object.defineProperty(ControlArray.prototype.at, "parameters", {get: function() {
    return [[assert.type.number]];
  }});
Object.defineProperty(ControlArray.prototype.push, "parameters", {get: function() {
    return [[AbstractControl]];
  }});
Object.defineProperty(ControlArray.prototype.insert, "parameters", {get: function() {
    return [[assert.type.number], [AbstractControl]];
  }});
Object.defineProperty(ControlArray.prototype.removeAt, "parameters", {get: function() {
    return [[assert.type.number]];
  }});
