// angular2/src/forms/validators.js
"use strict";
Object.defineProperties(module.exports, {
  Validators: {get: function() {
      return Validators;
    }},
  __esModule: {value: true}
});
var $__rtts_95_assert_47_rtts_95_assert__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_collection__,
    $__model__;
var assert = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__}).assert;
var $__1 = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}),
    isBlank = $__1.isBlank,
    isPresent = $__1.isPresent;
var $__2 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    List = $__2.List,
    ListWrapper = $__2.ListWrapper,
    StringMapWrapper = $__2.StringMapWrapper;
var modelModule = ($__model__ = require("./model"), $__model__ && $__model__.__esModule && $__model__ || {default: $__model__});
var Validators = function Validators() {
  ;
};
var $Validators = Validators;
($traceurRuntime.createClass)(Validators, {}, {
  required: function(c) {
    assert.argumentTypes(c, modelModule.Control);
    return isBlank(c.value) || c.value == "" ? {"required": true} : null;
  },
  nullValidator: function(c) {
    assert.argumentTypes(c, assert.type.any);
    return null;
  },
  compose: function(validators) {
    assert.argumentTypes(validators, assert.genericType(List, Function));
    return assert.returnType((function(c) {
      assert.argumentTypes(c, modelModule.Control);
      var res = ListWrapper.reduce(validators, (function(res, validator) {
        var errors = validator(c);
        return isPresent(errors) ? StringMapWrapper.merge(res, errors) : res;
      }), {});
      return StringMapWrapper.isEmpty(res) ? null : res;
    }), Function);
  },
  group: function(c) {
    assert.argumentTypes(c, modelModule.ControlGroup);
    var res = {};
    StringMapWrapper.forEach(c.controls, (function(control, name) {
      if (c.contains(name) && isPresent(control.errors)) {
        $Validators._mergeErrors(control, res);
      }
    }));
    return StringMapWrapper.isEmpty(res) ? null : res;
  },
  array: function(c) {
    assert.argumentTypes(c, modelModule.ControlArray);
    var res = {};
    ListWrapper.forEach(c.controls, (function(control) {
      if (isPresent(control.errors)) {
        $Validators._mergeErrors(control, res);
      }
    }));
    return StringMapWrapper.isEmpty(res) ? null : res;
  },
  _mergeErrors: function(control, res) {
    StringMapWrapper.forEach(control.errors, (function(value, error) {
      if (!StringMapWrapper.contains(res, error)) {
        res[error] = [];
      }
      ListWrapper.push(res[error], control);
    }));
  }
});
Object.defineProperty(Validators.required, "parameters", {get: function() {
    return [[modelModule.Control]];
  }});
Object.defineProperty(Validators.nullValidator, "parameters", {get: function() {
    return [[assert.type.any]];
  }});
Object.defineProperty(Validators.compose, "parameters", {get: function() {
    return [[assert.genericType(List, Function)]];
  }});
Object.defineProperty(Validators.group, "parameters", {get: function() {
    return [[modelModule.ControlGroup]];
  }});
Object.defineProperty(Validators.array, "parameters", {get: function() {
    return [[modelModule.ControlArray]];
  }});
