// angular2/src/core/annotations/annotations.js
"use strict";
Object.defineProperties(module.exports, {
  Directive: {get: function() {
      return Directive;
    }},
  Component: {get: function() {
      return Component;
    }},
  DynamicComponent: {get: function() {
      return DynamicComponent;
    }},
  Decorator: {get: function() {
      return Decorator;
    }},
  Viewport: {get: function() {
      return Viewport;
    }},
  onDestroy: {get: function() {
      return onDestroy;
    }},
  onChange: {get: function() {
      return onChange;
    }},
  onAllChangesDone: {get: function() {
      return onAllChangesDone;
    }},
  __esModule: {value: true}
});
var $__rtts_95_assert_47_rtts_95_assert__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_collection__,
    $__angular2_47_di__;
var assert = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__}).assert;
var $__1 = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}),
    ABSTRACT = $__1.ABSTRACT,
    CONST = $__1.CONST,
    normalizeBlank = $__1.normalizeBlank,
    isPresent = $__1.isPresent;
var $__2 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    ListWrapper = $__2.ListWrapper,
    List = $__2.List;
var Injectable = ($__angular2_47_di__ = require("angular2/di"), $__angular2_47_di__ && $__angular2_47_di__.__esModule && $__angular2_47_di__ || {default: $__angular2_47_di__}).Injectable;
var Directive = function Directive() {
  var $__5 = arguments[0] !== (void 0) ? arguments[0] : {},
      selector = $__5.selector,
      bind = $__5.bind,
      events = $__5.events,
      lifecycle = $__5.lifecycle;
  $traceurRuntime.superConstructor($Directive).call(this);
  this.selector = selector;
  this.bind = bind;
  this.events = events;
  this.lifecycle = lifecycle;
};
var $Directive = Directive;
($traceurRuntime.createClass)(Directive, {hasLifecycleHook: function(hook) {
    assert.argumentTypes(hook, assert.type.string);
    return assert.returnType((isPresent(this.lifecycle) ? ListWrapper.contains(this.lifecycle, hook) : false), assert.type.boolean);
  }}, {}, Injectable);
Object.defineProperty(Directive, "annotations", {get: function() {
    return [new ABSTRACT(), new CONST()];
  }});
Object.defineProperty(Directive.prototype.hasLifecycleHook, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
var Component = function Component() {
  var $__5 = arguments[0] !== (void 0) ? arguments[0] : {},
      selector = $__5.selector,
      bind = $__5.bind,
      events = $__5.events,
      services = $__5.services,
      lifecycle = $__5.lifecycle;
  $traceurRuntime.superConstructor($Component).call(this, {
    selector: selector,
    bind: bind,
    events: events,
    lifecycle: lifecycle
  });
  this.services = services;
};
var $Component = Component;
($traceurRuntime.createClass)(Component, {}, {}, Directive);
Object.defineProperty(Component, "annotations", {get: function() {
    return [new CONST()];
  }});
var DynamicComponent = function DynamicComponent() {
  var $__5 = arguments[0] !== (void 0) ? arguments[0] : {},
      selector = $__5.selector,
      bind = $__5.bind,
      events = $__5.events,
      services = $__5.services,
      lifecycle = $__5.lifecycle;
  $traceurRuntime.superConstructor($DynamicComponent).call(this, {
    selector: selector,
    bind: bind,
    events: events,
    lifecycle: lifecycle
  });
  this.services = services;
};
var $DynamicComponent = DynamicComponent;
($traceurRuntime.createClass)(DynamicComponent, {}, {}, Directive);
Object.defineProperty(DynamicComponent, "annotations", {get: function() {
    return [new CONST()];
  }});
var Decorator = function Decorator() {
  var $__6;
  var $__5 = arguments[0] !== (void 0) ? arguments[0] : {},
      selector = $__5.selector,
      bind = $__5.bind,
      events = $__5.events,
      lifecycle = $__5.lifecycle,
      compileChildren = ($__6 = $__5.compileChildren) === void 0 ? true : $__6;
  $traceurRuntime.superConstructor($Decorator).call(this, {
    selector: selector,
    bind: bind,
    events: events,
    lifecycle: lifecycle
  });
  this.compileChildren = compileChildren;
};
var $Decorator = Decorator;
($traceurRuntime.createClass)(Decorator, {}, {}, Directive);
Object.defineProperty(Decorator, "annotations", {get: function() {
    return [new CONST()];
  }});
var Viewport = function Viewport() {
  var $__5 = arguments[0] !== (void 0) ? arguments[0] : {},
      selector = $__5.selector,
      bind = $__5.bind,
      events = $__5.events,
      lifecycle = $__5.lifecycle;
  $traceurRuntime.superConstructor($Viewport).call(this, {
    selector: selector,
    bind: bind,
    events: events,
    lifecycle: lifecycle
  });
};
var $Viewport = Viewport;
($traceurRuntime.createClass)(Viewport, {}, {}, Directive);
Object.defineProperty(Viewport, "annotations", {get: function() {
    return [new CONST()];
  }});
var onDestroy = "onDestroy";
var onChange = "onChange";
var onAllChangesDone = "onAllChangesDone";
