// angular2/src/core/compiler/element_binder.js
"use strict";
Object.defineProperties(module.exports, {
  ElementBinder: {get: function() {
      return ElementBinder;
    }},
  __esModule: {value: true}
});
var $__rtts_95_assert_47_rtts_95_assert__,
    $__angular2_47_src_47_facade_47_lang__,
    $__element_95_injector__,
    $__directive_95_metadata__,
    $__angular2_47_src_47_facade_47_collection__,
    $__view__;
var assert = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__}).assert;
var $__1 = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}),
    int = $__1.int,
    isBlank = $__1.isBlank,
    BaseException = $__1.BaseException;
var eiModule = ($__element_95_injector__ = require("./element_injector"), $__element_95_injector__ && $__element_95_injector__.__esModule && $__element_95_injector__ || {default: $__element_95_injector__});
var DirectiveMetadata = ($__directive_95_metadata__ = require("./directive_metadata"), $__directive_95_metadata__ && $__directive_95_metadata__.__esModule && $__directive_95_metadata__ || {default: $__directive_95_metadata__}).DirectiveMetadata;
var $__3 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    List = $__3.List,
    StringMap = $__3.StringMap;
var viewModule = ($__view__ = require("./view"), $__view__ && $__view__.__esModule && $__view__ || {default: $__view__});
var ElementBinder = function ElementBinder(index, parent, distanceToParent, protoElementInjector, componentDirective, viewportDirective) {
  assert.argumentTypes(index, int, parent, $ElementBinder, distanceToParent, int, protoElementInjector, eiModule.ProtoElementInjector, componentDirective, DirectiveMetadata, viewportDirective, DirectiveMetadata);
  if (isBlank(index)) {
    throw new BaseException('null index not allowed.');
  }
  this.protoElementInjector = protoElementInjector;
  this.componentDirective = componentDirective;
  this.viewportDirective = viewportDirective;
  this.parent = parent;
  this.index = index;
  this.distanceToParent = distanceToParent;
  this.events = null;
  this.textNodeIndices = null;
  this.hasElementPropertyBindings = false;
  this.nestedProtoView = null;
  this.contentTagSelector = null;
};
var $ElementBinder = ElementBinder;
($traceurRuntime.createClass)(ElementBinder, {}, {});
Object.defineProperty(ElementBinder, "parameters", {get: function() {
    return [[int], [ElementBinder], [int], [eiModule.ProtoElementInjector], [DirectiveMetadata], [DirectiveMetadata]];
  }});
