// angular2/src/core/compiler/shadow_dom_emulation/content_tag.js
"use strict";
Object.defineProperties(module.exports, {
  Content: {get: function() {
      return Content;
    }},
  __esModule: {value: true}
});
var $__rtts_95_assert_47_rtts_95_assert__,
    $__light_95_dom__,
    $__angular2_47_di__,
    $__angular2_47_src_47_dom_47_dom_95_adapter__,
    $__angular2_47_src_47_facade_47_lang__,
    $__angular2_47_src_47_facade_47_collection__;
var assert = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__}).assert;
var ldModule = ($__light_95_dom__ = require("./light_dom"), $__light_95_dom__ && $__light_95_dom__.__esModule && $__light_95_dom__ || {default: $__light_95_dom__});
var $__1 = ($__angular2_47_di__ = require("angular2/di"), $__angular2_47_di__ && $__angular2_47_di__.__esModule && $__angular2_47_di__ || {default: $__angular2_47_di__}),
    Inject = $__1.Inject,
    Injectable = $__1.Injectable;
var DOM = ($__angular2_47_src_47_dom_47_dom_95_adapter__ = require("angular2/src/dom/dom_adapter"), $__angular2_47_src_47_dom_47_dom_95_adapter__ && $__angular2_47_src_47_dom_47_dom_95_adapter__.__esModule && $__angular2_47_src_47_dom_47_dom_95_adapter__ || {default: $__angular2_47_src_47_dom_47_dom_95_adapter__}).DOM;
var isPresent = ($__angular2_47_src_47_facade_47_lang__ = require("angular2/src/facade/lang"), $__angular2_47_src_47_facade_47_lang__ && $__angular2_47_src_47_facade_47_lang__.__esModule && $__angular2_47_src_47_facade_47_lang__ || {default: $__angular2_47_src_47_facade_47_lang__}).isPresent;
var $__4 = ($__angular2_47_src_47_facade_47_collection__ = require("angular2/src/facade/collection"), $__angular2_47_src_47_facade_47_collection__ && $__angular2_47_src_47_facade_47_collection__.__esModule && $__angular2_47_src_47_facade_47_collection__ || {default: $__angular2_47_src_47_facade_47_collection__}),
    List = $__4.List,
    ListWrapper = $__4.ListWrapper;
var ContentStrategy = function ContentStrategy() {
  ;
};
($traceurRuntime.createClass)(ContentStrategy, {insert: function(nodes) {
    assert.argumentTypes(nodes, List);
  }}, {});
Object.defineProperty(ContentStrategy.prototype.insert, "parameters", {get: function() {
    return [[List]];
  }});
var RenderedContent = function RenderedContent(contentEl) {
  $traceurRuntime.superConstructor($RenderedContent).call(this);
  this.beginScript = contentEl;
  this.endScript = DOM.nextSibling(this.beginScript);
  this.nodes = [];
};
var $RenderedContent = RenderedContent;
($traceurRuntime.createClass)(RenderedContent, {
  insert: function(nodes) {
    assert.argumentTypes(nodes, List);
    this.nodes = nodes;
    DOM.insertAllBefore(this.endScript, nodes);
    this._removeNodesUntil(ListWrapper.isEmpty(nodes) ? this.endScript : nodes[0]);
  },
  _removeNodesUntil: function(node) {
    var p = DOM.parentElement(this.beginScript);
    for (var next = DOM.nextSibling(this.beginScript); next !== node; next = DOM.nextSibling(this.beginScript)) {
      DOM.removeChild(p, next);
    }
  }
}, {}, ContentStrategy);
Object.defineProperty(RenderedContent, "annotations", {get: function() {
    return [new Injectable()];
  }});
Object.defineProperty(RenderedContent.prototype.insert, "parameters", {get: function() {
    return [[List]];
  }});
var IntermediateContent = function IntermediateContent(destinationLightDom) {
  assert.argumentTypes(destinationLightDom, ldModule.LightDom);
  $traceurRuntime.superConstructor($IntermediateContent).call(this);
  this.destinationLightDom = destinationLightDom;
  this.nodes = [];
};
var $IntermediateContent = IntermediateContent;
($traceurRuntime.createClass)(IntermediateContent, {insert: function(nodes) {
    assert.argumentTypes(nodes, List);
    this.nodes = nodes;
    this.destinationLightDom.redistribute();
  }}, {}, ContentStrategy);
Object.defineProperty(IntermediateContent, "parameters", {get: function() {
    return [[ldModule.LightDom]];
  }});
Object.defineProperty(IntermediateContent.prototype.insert, "parameters", {get: function() {
    return [[List]];
  }});
var Content = function Content(destinationLightDom, contentStartEl, selector) {
  assert.argumentTypes(destinationLightDom, ldModule.LightDom, contentStartEl, assert.type.any, selector, assert.type.string);
  this.select = selector;
  this.contentStartElement = contentStartEl;
  this._strategy = isPresent(destinationLightDom) ? new IntermediateContent(destinationLightDom) : new RenderedContent(contentStartEl);
};
($traceurRuntime.createClass)(Content, {
  nodes: function() {
    return assert.returnType((this._strategy.nodes), List);
  },
  insert: function(nodes) {
    assert.argumentTypes(nodes, List);
    this._strategy.insert(nodes);
  }
}, {});
Object.defineProperty(Content, "parameters", {get: function() {
    return [[ldModule.LightDom], [], [assert.type.string]];
  }});
Object.defineProperty(Content.prototype.insert, "parameters", {get: function() {
    return [[List]];
  }});
