// angular2/change_detection.js
"use strict";
var $__rtts_95_assert_47_rtts_95_assert__,
    $__src_47_change_95_detection_47_parser_47_ast__,
    $__src_47_change_95_detection_47_parser_47_lexer__,
    $__src_47_change_95_detection_47_parser_47_parser__,
    $__src_47_change_95_detection_47_parser_47_locals__,
    $__src_47_change_95_detection_47_exceptions__,
    $__src_47_change_95_detection_47_interfaces__,
    $__src_47_change_95_detection_47_proto_95_change_95_detector__,
    $__src_47_change_95_detection_47_dynamic_95_change_95_detector__,
    $__src_47_change_95_detection_47_binding_95_propagation_95_config__,
    $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__,
    $__src_47_change_95_detection_47_change_95_detection_95_util__,
    $__src_47_change_95_detection_47_pipes_47_pipe__,
    $__src_47_change_95_detection_47_proto_95_change_95_detector__,
    $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__,
    $__src_47_change_95_detection_47_pipes_47_iterable_95_changes__,
    $__src_47_change_95_detection_47_pipes_47_keyvalue_95_changes__,
    $__src_47_change_95_detection_47_pipes_47_null_95_pipe__;
var assert = ($__rtts_95_assert_47_rtts_95_assert__ = require("rtts_assert/rtts_assert"), $__rtts_95_assert_47_rtts_95_assert__ && $__rtts_95_assert_47_rtts_95_assert__.__esModule && $__rtts_95_assert_47_rtts_95_assert__ || {default: $__rtts_95_assert_47_rtts_95_assert__}).assert;
var $__src_47_change_95_detection_47_parser_47_ast__ = ($__src_47_change_95_detection_47_parser_47_ast__ = require("./src/change_detection/parser/ast"), $__src_47_change_95_detection_47_parser_47_ast__ && $__src_47_change_95_detection_47_parser_47_ast__.__esModule && $__src_47_change_95_detection_47_parser_47_ast__ || {default: $__src_47_change_95_detection_47_parser_47_ast__});
var $__src_47_change_95_detection_47_parser_47_lexer__ = ($__src_47_change_95_detection_47_parser_47_lexer__ = require("./src/change_detection/parser/lexer"), $__src_47_change_95_detection_47_parser_47_lexer__ && $__src_47_change_95_detection_47_parser_47_lexer__.__esModule && $__src_47_change_95_detection_47_parser_47_lexer__ || {default: $__src_47_change_95_detection_47_parser_47_lexer__});
var $__src_47_change_95_detection_47_parser_47_parser__ = ($__src_47_change_95_detection_47_parser_47_parser__ = require("./src/change_detection/parser/parser"), $__src_47_change_95_detection_47_parser_47_parser__ && $__src_47_change_95_detection_47_parser_47_parser__.__esModule && $__src_47_change_95_detection_47_parser_47_parser__ || {default: $__src_47_change_95_detection_47_parser_47_parser__});
var $__src_47_change_95_detection_47_parser_47_locals__ = ($__src_47_change_95_detection_47_parser_47_locals__ = require("./src/change_detection/parser/locals"), $__src_47_change_95_detection_47_parser_47_locals__ && $__src_47_change_95_detection_47_parser_47_locals__.__esModule && $__src_47_change_95_detection_47_parser_47_locals__ || {default: $__src_47_change_95_detection_47_parser_47_locals__});
var $__src_47_change_95_detection_47_exceptions__ = ($__src_47_change_95_detection_47_exceptions__ = require("./src/change_detection/exceptions"), $__src_47_change_95_detection_47_exceptions__ && $__src_47_change_95_detection_47_exceptions__.__esModule && $__src_47_change_95_detection_47_exceptions__ || {default: $__src_47_change_95_detection_47_exceptions__});
var $__src_47_change_95_detection_47_interfaces__ = ($__src_47_change_95_detection_47_interfaces__ = require("./src/change_detection/interfaces"), $__src_47_change_95_detection_47_interfaces__ && $__src_47_change_95_detection_47_interfaces__.__esModule && $__src_47_change_95_detection_47_interfaces__ || {default: $__src_47_change_95_detection_47_interfaces__});
var $__src_47_change_95_detection_47_proto_95_change_95_detector__ = ($__src_47_change_95_detection_47_proto_95_change_95_detector__ = require("./src/change_detection/proto_change_detector"), $__src_47_change_95_detection_47_proto_95_change_95_detector__ && $__src_47_change_95_detection_47_proto_95_change_95_detector__.__esModule && $__src_47_change_95_detection_47_proto_95_change_95_detector__ || {default: $__src_47_change_95_detection_47_proto_95_change_95_detector__});
var $__src_47_change_95_detection_47_dynamic_95_change_95_detector__ = ($__src_47_change_95_detection_47_dynamic_95_change_95_detector__ = require("./src/change_detection/dynamic_change_detector"), $__src_47_change_95_detection_47_dynamic_95_change_95_detector__ && $__src_47_change_95_detection_47_dynamic_95_change_95_detector__.__esModule && $__src_47_change_95_detection_47_dynamic_95_change_95_detector__ || {default: $__src_47_change_95_detection_47_dynamic_95_change_95_detector__});
var $__src_47_change_95_detection_47_binding_95_propagation_95_config__ = ($__src_47_change_95_detection_47_binding_95_propagation_95_config__ = require("./src/change_detection/binding_propagation_config"), $__src_47_change_95_detection_47_binding_95_propagation_95_config__ && $__src_47_change_95_detection_47_binding_95_propagation_95_config__.__esModule && $__src_47_change_95_detection_47_binding_95_propagation_95_config__ || {default: $__src_47_change_95_detection_47_binding_95_propagation_95_config__});
var $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__ = ($__src_47_change_95_detection_47_pipes_47_pipe_95_registry__ = require("./src/change_detection/pipes/pipe_registry"), $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__ && $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__.__esModule && $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__ || {default: $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__});
var $__src_47_change_95_detection_47_change_95_detection_95_util__ = ($__src_47_change_95_detection_47_change_95_detection_95_util__ = require("./src/change_detection/change_detection_util"), $__src_47_change_95_detection_47_change_95_detection_95_util__ && $__src_47_change_95_detection_47_change_95_detection_95_util__.__esModule && $__src_47_change_95_detection_47_change_95_detection_95_util__ || {default: $__src_47_change_95_detection_47_change_95_detection_95_util__});
var $__src_47_change_95_detection_47_pipes_47_pipe__ = ($__src_47_change_95_detection_47_pipes_47_pipe__ = require("./src/change_detection/pipes/pipe"), $__src_47_change_95_detection_47_pipes_47_pipe__ && $__src_47_change_95_detection_47_pipes_47_pipe__.__esModule && $__src_47_change_95_detection_47_pipes_47_pipe__ || {default: $__src_47_change_95_detection_47_pipes_47_pipe__});
var $__1 = ($__src_47_change_95_detection_47_proto_95_change_95_detector__ = require("./src/change_detection/proto_change_detector"), $__src_47_change_95_detection_47_proto_95_change_95_detector__ && $__src_47_change_95_detection_47_proto_95_change_95_detector__.__esModule && $__src_47_change_95_detection_47_proto_95_change_95_detector__ || {default: $__src_47_change_95_detection_47_proto_95_change_95_detector__}),
    ProtoChangeDetector = $__1.ProtoChangeDetector,
    DynamicProtoChangeDetector = $__1.DynamicProtoChangeDetector,
    JitProtoChangeDetector = $__1.JitProtoChangeDetector;
var PipeRegistry = ($__src_47_change_95_detection_47_pipes_47_pipe_95_registry__ = require("./src/change_detection/pipes/pipe_registry"), $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__ && $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__.__esModule && $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__ || {default: $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__}).PipeRegistry;
var IterableChangesFactory = ($__src_47_change_95_detection_47_pipes_47_iterable_95_changes__ = require("./src/change_detection/pipes/iterable_changes"), $__src_47_change_95_detection_47_pipes_47_iterable_95_changes__ && $__src_47_change_95_detection_47_pipes_47_iterable_95_changes__.__esModule && $__src_47_change_95_detection_47_pipes_47_iterable_95_changes__ || {default: $__src_47_change_95_detection_47_pipes_47_iterable_95_changes__}).IterableChangesFactory;
var KeyValueChangesFactory = ($__src_47_change_95_detection_47_pipes_47_keyvalue_95_changes__ = require("./src/change_detection/pipes/keyvalue_changes"), $__src_47_change_95_detection_47_pipes_47_keyvalue_95_changes__ && $__src_47_change_95_detection_47_pipes_47_keyvalue_95_changes__.__esModule && $__src_47_change_95_detection_47_pipes_47_keyvalue_95_changes__ || {default: $__src_47_change_95_detection_47_pipes_47_keyvalue_95_changes__}).KeyValueChangesFactory;
var NullPipeFactory = ($__src_47_change_95_detection_47_pipes_47_null_95_pipe__ = require("./src/change_detection/pipes/null_pipe"), $__src_47_change_95_detection_47_pipes_47_null_95_pipe__ && $__src_47_change_95_detection_47_pipes_47_null_95_pipe__.__esModule && $__src_47_change_95_detection_47_pipes_47_null_95_pipe__ || {default: $__src_47_change_95_detection_47_pipes_47_null_95_pipe__}).NullPipeFactory;
var ChangeDetection = function ChangeDetection() {
  ;
};
($traceurRuntime.createClass)(ChangeDetection, {createProtoChangeDetector: function(name) {
    assert.argumentTypes(name, assert.type.string);
    return assert.returnType((null), ProtoChangeDetector);
  }}, {});
Object.defineProperty(ChangeDetection.prototype.createProtoChangeDetector, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
var defaultPipes = {
  "iterableDiff": [new IterableChangesFactory(), new NullPipeFactory()],
  "keyValDiff": [new KeyValueChangesFactory(), new NullPipeFactory()]
};
var DynamicChangeDetection = function DynamicChangeDetection(registry) {
  assert.argumentTypes(registry, PipeRegistry);
  $traceurRuntime.superConstructor($DynamicChangeDetection).call(this);
  this.registry = registry;
};
var $DynamicChangeDetection = DynamicChangeDetection;
($traceurRuntime.createClass)(DynamicChangeDetection, {createProtoChangeDetector: function(name) {
    assert.argumentTypes(name, assert.type.string);
    return assert.returnType((new DynamicProtoChangeDetector(this.registry)), ProtoChangeDetector);
  }}, {}, ChangeDetection);
Object.defineProperty(DynamicChangeDetection, "parameters", {get: function() {
    return [[PipeRegistry]];
  }});
Object.defineProperty(DynamicChangeDetection.prototype.createProtoChangeDetector, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
var JitChangeDetection = function JitChangeDetection(registry) {
  assert.argumentTypes(registry, PipeRegistry);
  $traceurRuntime.superConstructor($JitChangeDetection).call(this);
  this.registry = registry;
};
var $JitChangeDetection = JitChangeDetection;
($traceurRuntime.createClass)(JitChangeDetection, {createProtoChangeDetector: function(name) {
    assert.argumentTypes(name, assert.type.string);
    return assert.returnType((new JitProtoChangeDetector(this.registry)), ProtoChangeDetector);
  }}, {}, ChangeDetection);
Object.defineProperty(JitChangeDetection, "parameters", {get: function() {
    return [[PipeRegistry]];
  }});
Object.defineProperty(JitChangeDetection.prototype.createProtoChangeDetector, "parameters", {get: function() {
    return [[assert.type.string]];
  }});
var _registry = new PipeRegistry(defaultPipes);
var dynamicChangeDetection = new DynamicChangeDetection(_registry);
var jitChangeDetection = new JitChangeDetection(_registry);
module.exports = $traceurRuntime.exportStar({
  get AST() {
    return $__src_47_change_95_detection_47_parser_47_ast__.AST;
  },
  get Lexer() {
    return $__src_47_change_95_detection_47_parser_47_lexer__.Lexer;
  },
  get Parser() {
    return $__src_47_change_95_detection_47_parser_47_parser__.Parser;
  },
  get Locals() {
    return $__src_47_change_95_detection_47_parser_47_locals__.Locals;
  },
  get ExpressionChangedAfterItHasBeenChecked() {
    return $__src_47_change_95_detection_47_exceptions__.ExpressionChangedAfterItHasBeenChecked;
  },
  get ChangeDetectionError() {
    return $__src_47_change_95_detection_47_exceptions__.ChangeDetectionError;
  },
  get ChangeRecord() {
    return $__src_47_change_95_detection_47_interfaces__.ChangeRecord;
  },
  get ChangeDispatcher() {
    return $__src_47_change_95_detection_47_interfaces__.ChangeDispatcher;
  },
  get ChangeDetector() {
    return $__src_47_change_95_detection_47_interfaces__.ChangeDetector;
  },
  get CHECK_ONCE() {
    return $__src_47_change_95_detection_47_interfaces__.CHECK_ONCE;
  },
  get CHECK_ALWAYS() {
    return $__src_47_change_95_detection_47_interfaces__.CHECK_ALWAYS;
  },
  get DETACHED() {
    return $__src_47_change_95_detection_47_interfaces__.DETACHED;
  },
  get CHECKED() {
    return $__src_47_change_95_detection_47_interfaces__.CHECKED;
  },
  get ProtoChangeDetector() {
    return $__src_47_change_95_detection_47_proto_95_change_95_detector__.ProtoChangeDetector;
  },
  get DynamicProtoChangeDetector() {
    return $__src_47_change_95_detection_47_proto_95_change_95_detector__.DynamicProtoChangeDetector;
  },
  get JitProtoChangeDetector() {
    return $__src_47_change_95_detection_47_proto_95_change_95_detector__.JitProtoChangeDetector;
  },
  get BindingRecord() {
    return $__src_47_change_95_detection_47_proto_95_change_95_detector__.BindingRecord;
  },
  get DynamicChangeDetector() {
    return $__src_47_change_95_detection_47_dynamic_95_change_95_detector__.DynamicChangeDetector;
  },
  get BindingPropagationConfig() {
    return $__src_47_change_95_detection_47_binding_95_propagation_95_config__.BindingPropagationConfig;
  },
  get uninitialized() {
    return $__src_47_change_95_detection_47_change_95_detection_95_util__.uninitialized;
  },
  get ChangeDetection() {
    return ChangeDetection;
  },
  get defaultPipes() {
    return defaultPipes;
  },
  get DynamicChangeDetection() {
    return DynamicChangeDetection;
  },
  get JitChangeDetection() {
    return JitChangeDetection;
  },
  get dynamicChangeDetection() {
    return dynamicChangeDetection;
  },
  get jitChangeDetection() {
    return jitChangeDetection;
  },
  __esModule: true
}, $__src_47_change_95_detection_47_pipes_47_pipe_95_registry__, $__src_47_change_95_detection_47_pipes_47_pipe__);
